### ISSUE TITLE
Create Sprite Classification Model

### ISSUE THREAD

**Issue by flyxiv** [flyxiv/nino#15]
Train model that classifies each sprite (move left, move downright, idle right, ...)

**Comment by flyxiv** [comment]
```python
SPRITE_DIRECTIONS = ['up', 'upright', 'right', 'downright', 'down', 'downleft', 'left', 'upleft']
SPRITE_DIRECTION_MOTIONS = ['idle', 'walk', 'run', 'attack', 'laydown', 'swim', 'jump']
SPRITE_NO_DIRECTION_MOTIONS = ['climb']

import os
import argparse

def parse_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument('--outdir', type=str, required=True, help='directory to create sprite folders')

    return parser.parse_args()

def create_sprite_folders(outdir):
    for direction in SPRITE_DIRECTIONS:
        for motion in SPRITE_DIRECTION_MOTIONS:
            folder_name = f'{motion}_{direction}'
            print(f'Creating folder: {folder_name}')

            if not os.path.exists(os.path.join(outdir, folder_name)):
                os.makedirs(os.path.join(outdir, folder_name), exist_ok=True)

    for motion in SPRITE_NO_DIRECTION_MOTIONS:
        folder_name = f'{motion}'
        if not os.path.exists(os.path.join(outdir, folder_name)):
            os.makedirs(os.path.join(outdir, folder_name), exist_ok=True)

if __name__ == '__main__':
    args = parse_arguments()
    create_sprite_folders(args.outdir)
```

created folder for each sprite

**Comment by flyxiv** [comment]
Use hand labeled dataset in COCO format to generate data.

![Image](https://github.com/user-attachments/assets/c3ce73b3-0af0-4c0f-a9ef-c01246b95cf7)

**Comment by flyxiv** [comment]
```python
       for img_id in coco.getImgIds():
            img_info = coco.loadImgs(img_id)[0]
            img = cv2.imread(os.path.join(image_dir, img_info['file_name']))

            # Get annotation
            ann_ids = coco.getAnnIds(img_id)
            anns = coco.loadAnns(ann_ids)

            for ann_idx, ann in enumerate(anns):
                # Get sprite classification label
                sprite_classification_label = coco_labels[ann['category_id']]

                # Extract sprites
                rles = mask_utils.frPyObjects(ann['segmentation'], img_info['height'], img_info['width'])

                mask = mask_utils.decode(rles)
                mask = cv2.resize(mask, (img.shape[1], img.shape[0]))

                # 마스크 적용 (마스크=1: 원본 이미지, 마스크=0: 흰색 배경)
                white_background = np.ones_like(img) * 255
                masked_crop = np.zeros_like(img)
                
                for c in range(3):  # RGB 채널
                    masked_crop[:, :, c] = np.where(mask == 1, 
                                                    img[:, :, c], 
                                                    white_background[:, :, c])

                x1, y1, w, h = ann['bbox']

                x1 = max(0, int(x1))
                y1 = max(0, int(y1))
                x2 = min(img.shape[1], int(x1 + w))
                y2 = min(img.shape[0], int(y1 + h))

                sprite = masked_crop[y1:y2, x1:x2].copy()

                img_save_path = os.path.join(output_dir, sprite_classification_label)

                if not os.path.isdir(img_save_path):
                    os.makedirs(img_save_path, exist_ok=True)

                cv2.imwrite(os.path.join(img_save_path, f'{img_info["file_name"]}_{ann_idx}.png'), sprite)
                print(f"save {img_info['file_name']}_{ann_idx}.png in directory {img_save_path}")
```

create sprite instances from coco segment and bbox annotations

**Comment by flyxiv** [comment]
![Image](https://github.com/user-attachments/assets/77814ea4-036c-4d5f-9fb1-a108ed40b7d7)

all data were loaded nicely

**Comment by flyxiv** [comment]
Create split datasets

```python
import torch
from dataclasses import dataclass
from torch.utils.data import Dataset
from util.sprite_classifications import SPRITE_IDS 

import os
from sklearn.model_selection import train_test_split

@dataclass
class SpriteClassificationData:
    image_path: str
    label: int 

class SpriteClassificationDataTable():
    def __init__(self, data_dir: str):
        self.classification_data = list()

        for sprite_name in SPRITE_IDS.keys():
            for file in os.listdir(os.path.join(data_dir, sprite_name)):
                label = SPRITE_IDS[sprite_name]
                if file.endswith('.png') or file.endswith('.jpg'):
                    self.classification_data.append(SpriteClassificationData(os.path.join(data_dir, sprite_name, file), label))
    
    def split_to_train_val_test(self, train_ratio: float = 0.8, val_ratio: float = 0.1, test_ratio: float = 0.1):
        train_data, val_test_data = train_test_split(self.classification_data, test_size=test_ratio, random_state=42)
        val_data, test_data = train_test_split(val_test_data, test_size=val_ratio / (val_ratio + test_ratio), random_state=42)

        train_dataset = SpriteClassificationDataset(train_data)
        val_dataset = SpriteClassificationDataset(val_data)
        test_dataset = SpriteClassificationDataset(test_data)

        return train_dataset, val_dataset, test_dataset


class SpriteClassificationDataset(Dataset):
    def __init__(self, data: list[SpriteClassificationData]):
        self.data = data

    def __len__(self):
        return len(self.data)

    def __getitem__(self, index):
        image_path, label = self.data[index]

        image = Image.open(image_path)

        return image, label
```

**Comment by flyxiv** [comment]
Implemented multi classification for fine tuning EfficientNet

```python
import torch
import torch.nn as nn
import torchvision
import argparse
from torch.utils.data import DataLoader
from torch.optim import AdamW
from torchvision import transforms
from model.sprite_classification.dataset import SpriteClassificationDataTable
from util.sprite_classifications import SPRITE_IDS
from tqdm import tqdm

SPRITE_IMG_SIZE = (80, 40)

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('--data-dir', type=str, required=True, help='Path to the data directory')
    parser.add_argument('--batch-size', type=int, default=16, help='Batch size')
    parser.add_argument('--epochs', type=int, default=200, help='Number of epochs')
    parser.add_argument('--lr', type=float, default=0.001, help='Learning rate')
    return parser.parse_args()

if __name__ == '__main__':
    args = parse_args()
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Resize(SPRITE_IMG_SIZE),
    ])

    train_dataset, val_dataset, test_dataset = SpriteClassificationDataTable(args.data_dir, transform).split_to_train_val_test()

    train_loader = DataLoader(train_dataset, batch_size=args.batch_size, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=args.batch_size, shuffle=False)
    test_loader = DataLoader(test_dataset, batch_size=args.batch_size, shuffle=False)

    model = torchvision.models.efficientnet_b6(weights=True)
    model.classifier[-1] = nn.Linear(model.classifier[-1].in_features, len(SPRITE_IDS))
    model.to(device)

    criterion = nn.CrossEntropyLoss()
    optimizer = AdamW(model.parameters(), lr=args.lr)

    for epoch in tqdm(range(args.epochs)):
        model.train()
        running_loss = 0.0

        for (images, labels) in train_loader:
            images = images.to(device)
            labels = labels.to(device)

            optimizer.zero_grad()
            outputs = model(images)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()
        
        print(loss)

        model.eval()

        false_positive_per_class = torch.zeros(len(SPRITE_IDS))
        false_negative_per_class = torch.zeros(len(SPRITE_IDS))
        true_positive_per_class = torch.zeros(len(SPRITE_IDS))

        with torch.no_grad():
            for images, labels in tqdm(val_loader):
                images = images.to(device)
                labels = labels.to(device)

                outputs = model(images)
                _, pred_classes = outputs.max(1)

                true_positive_class_indices = torch.where(pred_classes == labels, pred_classes, -1) 
                false_positive_class_indices = torch.where(pred_classes != labels, pred_classes, -1)
                false_negative_class_indices = torch.where(pred_classes != labels, labels, -1)

                true_positive_class_indices = true_positive_class_indices[true_positive_class_indices != -1].to('cpu')
                false_positive_class_indices = false_positive_class_indices[false_positive_class_indices != -1].to('cpu')
                false_negative_class_indices = false_negative_class_indices[false_negative_class_indices != -1].to('cpu')

                true_positive_per_class[true_positive_class_indices] += 1
                false_positive_per_class[false_positive_class_indices] += 1
                false_negative_per_class[false_negative_class_indices] += 1


        print(f"Epoch {epoch + 1}, Precision: {true_positive_per_class.sum() / (true_positive_per_class + false_positive_per_class).sum()}")
        print(f"Epoch {epoch + 1}, Recall: {true_positive_per_class.sum() / (true_positive_per_class.sum() + false_negative_per_class.sum())}")
```

**Comment by flyxiv** [comment]
running optuna experiments to tune hyperparameters.

**Comment by flyxiv** [comment]
Classification model with Efficientnet showing only 60% precision/recall

Going to try a different model

**Comment by flyxiv** [comment]
There was a bug in the dataset. Fixing the bug makes the model achieve 100% precision/recall on validation set.

![Image](https://github.com/user-attachments/assets/0f53d79d-168b-44f8-b583-0e753122961d)

**Comment by flyxiv** [comment]
Implement this model to deployment for the final step.
